Move the code that deploys a SageMaker Endpoint to a plugin that implements BaseDeploymentClient

jamestran201 opened the pull request:
Signed-off-by: James Tran <tran.james2001@gmail.com>

## What changes are proposed in this pull request?

This PR adds a deployment plugin for SageMaker called `SageMakerDeploymentClient`. Deployments to SageMaker can now be managed through the `mlflow deployments` CLI. The target URI for this command can be:
- ``sagemaker``: This will set the default region to `us-west-2` and the default assumed role ARN to `None`.
- ``sagemaker:/region_name``: This will set the default region to `region_name` and the default assumed role ARN to `None`.
- ``sagemaker:/region_name/assumed_role_arn``: This will set the default region to `region_name` and the default assumed role ARN to `assumed_role_arn`.

For example, to create a SageMaker deployment: `mlflow deployments create --target sagemaker:/us-east-1/arn:aws:iam::123456789012:role/assumed_role --name dummy --model-uri ...`

Related issue: https://github.com/mlflow/mlflow/issues/4816

## How is this patch tested?

This patch is tested by the tests in `tests/sagemaker/test_sagemaker_deployment_client.py`.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

Add a deployment plugin for SageMaker. The only implemented function at the moment is `create_deployment()`.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [x] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [x] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

jamestran201-alt committed:
Move the code that deploys a SageMaker Endpoint to a plugin that implements BaseDeploymentClient

Signed-off-by: James Tran <tmnhat2001@gmail.com>

tmnhat2001 commented on docs/source/conf.py:
```diff
@@ -331,6 +331,7 @@
     ("py:class", "matplotlib.figure.Figure"),
     ("py:class", "plotly.graph_objects.Figure"),
     ("py:class", "PIL.Image.Image"),
+    ("py:class", "mlflow.deployments.base.BaseDeploymentClient"),
```
This is to avoid this error when building the documentation:
```
/home/teppi/mlflow/mlflow/sagemaker/__init__.py:docstring of mlflow.sagemaker.SageMakerPlugin:1:py:class reference target not found: mlflow.deployments.base.BaseDeploymentClient
```

tmnhat2001 replied:
I believe it comes from this line: https://github.com/mlflow/mlflow/blob/bacc5d5aa903a9c9484ca75c218818159f8e641c/mlflow/sagemaker/__init__.py#L1748

The error goes away if BaseDeploymentClient is not declared as the parent class.

dbczumar replied:
Got it! Thanks for clarifying :)

tmnhat2001 commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1656,6 +1657,182 @@ def _does_model_exist(model_name, sage_client):
+
+        return config
+
+    @experimental
+    def create_deployment(self, name, model_uri, flavor=None, config=None):
```
Marking this as experimental because not all methods in this class have been implemented yet.

tmnhat2001 commented on tests/sagemaker/test_sagemaker_plugin.py:
```diff
@@ -0,0 +1,615 @@
+import os
```
Most of the content in this file are copied from https://github.com/mlflow/mlflow/blob/master/tests/sagemaker/test_deployment.py.

The only changes are:
- Call `sagemaker_plugin.create_deployment()` instead of `mfs.deploy()`
- Create a fixture for the SageMakerPlugin

tmnhat2001 commented on tests/sagemaker/test_sagemaker_plugin.py:
```diff
@@ -0,0 +1,615 @@
+
+# @pytest.mark.large
+# @mock_sagemaker_aws_services
+# def test_deploy_replace_in_asynchronous_mode_returns_before_endpoint_creation_completes(
```
Some tests are commented out in this file because they are testing the ability to update a deployment. This functionality will be handled by `update_deployment()` in a follow-up PR.

dbczumar reviewed (commented):
@tmnhat2001 Apologies for the delay. This looks like a great start for the implementation! A discussed on the issue, can we explore embedding the region name in the target URI?

jamestran201-alt committed:
Merge branch 'mlflow:master' into sagemaker-migration-1

jamestran201-alt committed:
Set default AWS region through target_uri

Signed-off-by: James Tran <tmnhat2001@gmail.com>

tmnhat2001 commented:
@dbczumar No worries! I was leaving it for a later PR, but I've added them here now. Please review again when you have the chance.

dbczumar commented:
@tmnhat2001 Thank you for your updates - I'll take a look first thing tomorrow morning.

dbczumar commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1656,6 +1659,204 @@ def _does_model_exist(model_name, sage_client):
+
+    :param target_uri: A URI that has the value of ``sagemaker`` or ``sagemaker:/region_name``.
+                       The provided region name will be used as the default region for subsequent function calls.
+                       When no region name is provided, the default region will be set to `us-west-2`.
```
Nice! Apologies for continuing to make things harder, but we recently merged https://github.com/mlflow/mlflow/pull/4923 to enable users to specify an IAM role to use when interfacing with SageMaker deployments, which makes it easier to manage deployments across accounts. It would be great to incorporate IAM roles into the target URI to support this use case as well.

Perhaps `sagemaker:/<region_name>/<iam_role_arn>`? Users may have to URL encode the role ARN if it contains `?` or `#`, but I think this is probably uncommon.

If an IAM role ARN is specified, the user must also specify the region. Let me know how this sounds.

tmnhat2001 replied:
That sounds great!

dbczumar reviewed (commented):
@tmnhat2001 This is looking great! Left a few comments. Once those are addressed, I think we can move forward with implementing the other deployment client methods. Thanks so much for your contribution!

jamestran201-alt committed:
Merge branch 'mlflow:master' into sagemaker-migration-1

tmnhat2001 commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1656,6 +1659,204 @@ def _does_model_exist(model_name, sage_client):
+            timeout_seconds=config["timeout_seconds"],
+        )
+
+    def update_deployment(self, name, model_uri=None, flavor=None, config=None):
```
Yes, that's right.

jamestran201-alt committed:
Rename SageMaker deployment client

Signed-off-by: James Tran <tmnhat2001@gmail.com>

jamestran201-alt committed:
Parse assumed role ARN from target URI

Signed-off-by: James Tran <tmnhat2001@gmail.com>

jamestran201-alt committed:
Add `_target_help()` to follow the deployment plugin interface

Signed-off-by: James Tran <tmnhat2001@gmail.com>

jamestran201-alt committed:
Register SageMakerDeploymentClient with the deployment plugin manager

Signed-off-by: James Tran <tmnhat2001@gmail.com>

tmnhat2001 commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1656,6 +1659,204 @@ def _does_model_exist(model_name, sage_client):
         return True if response else False
 
 
+class SageMakerPlugin(BaseDeploymentClient):
```
I had to register the module that contains the SageMaker client (see https://github.com/mlflow/mlflow/pull/4971/commits/059d64f3f77f17f802da3692a6d16a7e5a6c50c4) because the plugin manager requires some functions are defined at the module level when initializing a client object:
https://github.com/mlflow/mlflow/blob/9faf3b04c9f6b3766a64d7433dd3e85bcdf27fff/mlflow/deployments/plugin_manager.py#L93-L94

tmnhat2001 commented:
@dbczumar Thanks for the review! I have addressed the comments. Please review it again when you have time.

jamestran201-alt committed:
Merge remote-tracking branch 'upstream/master' into sagemaker-migration-1

 Conflicts:
	mlflow/sagemaker/__init__.py

dbczumar commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1064,6 +1066,33 @@ def _sigterm_handler(*_):
     proc.wait()
 
 
+def target_help(_):
```
Can we remove the `_` arg from `target_help()` to resolve the following issue?:

```
$ mlflow deployments help -t sagemaker
/Users/corey.zumar/mlflow/mlflow/server/handlers.py:119: UserWarning: Failure attempting to register store for scheme "file-plugin": No module named 'mlflow_test_plugin.sqlalchemy_store'
  self.register_entrypoints()
<module 'mlflow.sagemaker' from '/Users/corey.zumar/mlflow/mlflow/sagemaker/__init__.py'>
Traceback (most recent call last):
  File "/Users/corey.zumar/opt/anaconda3/bin/mlflow", line 33, in <module>
    sys.exit(load_entry_point('mlflow', 'console_scripts', 'mlflow')())
  File "/Users/corey.zumar/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/corey.zumar/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/corey.zumar/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/corey.zumar/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/corey.zumar/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/corey.zumar/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/corey.zumar/mlflow/mlflow/deployments/cli.py", line 216, in target_help
    click.echo(interface._target_help(target))
  File "/Users/corey.zumar/mlflow/mlflow/deployments/interface.py", line 83, in _target_help
    return plugin_store[target].target_help()
TypeError: target_help() missing 1 required positional argument: '_'
```

dbczumar commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1064,6 +1066,33 @@ def _sigterm_handler(*_):
+    The `create` and `update` commands require a deployment name and a model_uri. The model flavor
+    and deployment configuration can be optionally provided.
+
+    The `delete` command accepts configurations to archive a model instead of deleting, execute
```
Awesome! To support this, we'll need to extend the interface to accomplish this - e.g. by adding `config` as an optional kwarg to `delete_deployment`.

dbczumar commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1064,6 +1066,33 @@ def _sigterm_handler(*_):
+    When the region_name or assume_role_arn are provided, they will be used as the default region
+    and assumed role ARN when executing the commands.
+
+    The `create` and `update` commands require a deployment name and a model_uri. The model flavor
+    and deployment configuration can be optionally provided.
```
Nice docstring! Can we also clarify that sync & async execution are supported, since we highlight that for `delete` below as well?

dbczumar commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1747,6 +1776,238 @@ def _does_model_exist(model_name, sage_client):
+        else:
+            # values_str would look like us-east-1/arn:aws:1234:role/assumed_role
+            self.region_name = values_str[:separator_index]
+            self.assumed_role_arn = values_str[separator_index + 1 :]
```
Can we handle the case where there may be multiple interior slashes in a path? e.g. `sagemaker:/us-west-2////arn:iam:my_role`?

dbczumar commented on tests/sagemaker/test_sagemaker_deployment_client.py:
```diff
@@ -0,0 +1,722 @@
+
+
+# @pytest.mark.large
+# @mock_sagemaker_aws_services
```
Let me know if I can help debug the commented tests or if they're a work in progress or just used for debugging during your implementation.

dbczumar reviewed (commented):
@tmnhat2001 This looks fantastic! Sorry for the delay here again - the holiday season has been quite busy. Barring a few comments, I think we're all set to implement the remaining APIs.

tmnhat2001 commented:
@dbczumar Sorry for the late reply. I need to wrap up a couple of things so I'll get back to this PR later this week.

jamestran201-alt committed:
Merge branch 'mlflow:master' into sagemaker-migration-1

jamestran201-alt committed:
Fix invalid argument and update docstring

Signed-off-by: James Tran <tmnhat2001@gmail.com>

jamestran201-alt committed:
Handle the case when there are interior slashes in the target_uri for Sagemaker deployment client

Signed-off-by: James Tran <tmnhat2001@gmail.com>

tmnhat2001 commented:
@dbczumar Sorry for the late response 🙇‍♂️ . I've addressed the comments, please review again when you have time.

Should I implement the remaining API's in this PR as well?

dbczumar commented:
> @dbczumar Sorry for the late response 🙇‍♂️ . I've addressed the comments, please review again when you have time.
> 
> Should I implement the remaining API's in this PR as well?

Hi @tmnhat2001 , no problem at all! If possible, it would be great to implement the remaining APIs in this PR :). Thank you for your contribution!

jamestran201-alt committed:
Change SageMaker default deployment config to account for both creating a new deployment and updating an existing one

Signed-off-by: James Tran <tmnhat2001@gmail.com>

jamestran201-alt committed:
Return model name and flavor after completing deployment to Sagemaker. Copy over missing tests

Signed-off-by: James Tran <tmnhat2001@gmail.com>

jamestran201-alt committed:
Add another test for creating a deployment through the CLI

Signed-off-by: James Tran <tmnhat2001@gmail.com>

jamestran201-alt committed:
Add function to update SageMaker deployment

Signed-off-by: James Tran <tmnhat2001@gmail.com>

jamestran201-alt committed:
Fix docs

Signed-off-by: James Tran <tmnhat2001@gmail.com>

jamestran201-alt committed:
Merge branch 'mlflow:master' into sagemaker-migration-1

tmnhat2001 commented:
@dbczumar Could you please review again? I have implemented the `update_deployment()` function.

jamestran201-alt committed:
Add `config` as an optional argument to the deployment plugin interface.
This is so that we can support deleting SageMaker endpoints.

The deployment CLI now accepts `config` as a flag argument. The config arguments will
only be passed to the `delete_deployment()` function when the function accepts it
as an argument. Otherwise, the config arguments will be omitted to
maintain backwards compatibility with the existing plugin implementations.

Signed-off-by: James Tran <tmnhat2001@gmail.com>

jamestran201 committed:
Merge branch 'sagemaker-migration-1' of github.com:jamestran201/mlflow into sagemaker-migration-1

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 committed:
Implement `delete_deployment()` function for SageMaker plugin

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 commented:
The commits between this comment and the previous one are for implementing the `delete_deployment()` function.

jamestran201 committed:
Implement `get_deployment()` function for SageMaker plugin

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 committed:
Address duplication in tests

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 commented:
The commits between this comment and the previous one are for implementing the `get_deployment()` function.

jamestran201 committed:
Implement `list_deployments()` function for SageMaker plugin

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 commented:
@dbczumar The `list_deployments()` function has been implemented. Should we implement the `predict()` function as well?

dbczumar commented:
> @dbczumar The `list_deployments()` function has been implemented. Should we implement the `predict()` function as well?

@jamestran201 Fantastic! Thanks so much! I'll take a look! Perhaps we can implement `predict()` as a separate PR since we don't currently have an equivalent function and it might require some more iterations. Would you be able to work on this in a separate PR?

jamestran201 committed:
Display the correct model flavor in the CLI output of create and update commands. Use the region name provided in target_uri as default when deleting deployment.

Signed-off-by: James Tran <tran.james2001@gmail.com>

dbczumar commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1747,6 +1779,480 @@ def _does_model_exist(model_name, sage_client):
+            instance_type=final_config["instance_type"],
+            instance_count=final_config["instance_count"],
+            vpc_config=final_config["vpc_config"],
+            synchronous=final_config["synchronous"],
```
From the command line, `synchronous` is being parsed as a String and always evaluates to `True` because the string value `"False"` is truthy. Can we fix this here and in the other methods that use synchronous (e.g. perhaps by adding special case logic to `apply_custom_config`)?

We could define a variable above as:

```
synchronous = final_config["synchronous"]
if isinstance(synchronous, str):
    synchronous = (synchronous == True)
```

and pass it in here (though we may want to do this in `apply_custom_config` to avoid code repetition):

```suggestion
            synchronous=synchronous,
```

dbczumar replied:
The same is true for `archive`. Can we update it as well?

dbczumar commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1747,6 +1779,480 @@ def _does_model_exist(model_name, sage_client):
+            bucket=final_config["bucket"],
+            image_url=final_config["image_url"],
+            region_name=final_config["region_name"],
+            mode=DEPLOYMENT_MODE_CREATE,
```
On second thought, can we check if `mode` is specified in the `config` and, if it is, pass it in? If `mode` is not specified, we can fall back to `DEPLOYMENT_MODE_CREATE`.

There are a couple reasons I think we should do this:

1. Right now, if I try to use this method while specifying a deployment name that already exists, I get the following error message:

```
mlflow.exceptions.MlflowException: You are attempting to deploy an application with name: mydep2 in 'create' mode. However, an application with the same name already exists. If you want to update this application, deploy in 'add' or 'replace' mode.
```

This exception comes from `mlflow.sagemaker.deploy()` and would be inaccurate if the `replace` and `add` `mode`s couldn't be specified.

2. For users who are automating deployments and don't want to add logic to check whether or not a deployment exists before creating a new one, supporting `mode=replace` would provide users with a convenient single-method "create or update" capability.

dbczumar commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1747,6 +1779,480 @@ def _does_model_exist(model_name, sage_client):
+            )
+
+    def predict(self, deployment_name, df):
+        pass
+
+    def explain(self, deployment_name, df):  # pylint: disable=unused-argument
+        pass
```
Can we raise `NotImplementedError` for now?

dbczumar commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1747,6 +1779,480 @@ def _does_model_exist(model_name, sage_client):
+            instance_count=final_config["instance_count"],
+            vpc_config=final_config["vpc_config"],
+            synchronous=final_config["synchronous"],
+            timeout_seconds=final_config["timeout_seconds"],
```
attempting to use this via the command line yields:

```
TypeError: '<' not supported between instances of 'float' and 'str'
```

Can we cast to `int()`?

dbczumar commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1747,6 +1779,480 @@ def _does_model_exist(model_name, sage_client):
+            mode=DEPLOYMENT_MODE_CREATE,
+            archive=final_config["archive"],
+            instance_type=final_config["instance_type"],
+            instance_count=final_config["instance_count"],
```
When instance count is passed via the command line, it should still be interpreted as an int, e.g.

```suggestion
            instance_count=int(final_config["instance_count"]),
```

Perhaps it would be better to make this change in `apply_custom_config`.

dbczumar commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1747,6 +1779,480 @@ def _does_model_exist(model_name, sage_client):
+            archive=final_config["archive"],
+            instance_type=final_config["instance_type"],
+            instance_count=final_config["instance_count"],
+            vpc_config=final_config["vpc_config"],
```
If `vpc_config` is passed via the commandline, its type will be `str`. In that case, can we parse it to a dictionary via `json.parse()`?

dbczumar reviewed (commented):
@jamestran201 Thank you so much for your hard work on this. The plugin works extremely well; I've tested out all of the CRUD functionality, and it's very smooth in general. I've left a few small comments, mostly related to docs and data type casting when using the tool through the `mlflow deployments` CLI. After these addressed, I'd imagine we'll be ready or very close to ready to merge. Let me know if you have any questions!

jamestran201 committed:
Use `assumed_role_arn` in `list_deployments()` when it is provided

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 committed:
Use `assumed_role_arn` in `get_deployment()` when it is specified

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 committed:
Move `None` check into the `_assume_role_and_get_credentials()` function to reduce duplication

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 committed:
Raise NotImplementedError for functions that are not implemented yet

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 committed:
Add examples for `delete_deployment()`

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 committed:
Convert some configs into correct types when they are passed through the CLI

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1747,6 +1769,574 @@ def _does_model_exist(model_name, sage_client):
+            if key not in config:
+                continue
+
+            if key in int_fields and not isinstance(value, int):
+                value = int(value)
+            elif key in bool_fields and not isinstance(value, bool):
+                value = value == "True"
+            elif key in dict_fields and not isinstance(value, dict):
+                value = json.parse(value)
```
There is some duplication in the conditionals. I attempted to DRY it up but the result seemed harder to follow. Please let me know if the duplication needs to be addressed.

dbczumar replied:
Love it! Looks great!

jamestran201 committed:
Fix incorrect docstring

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 committed:
Add usage examples to `update_deployment()`

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 committed:
Fix incorrect function name and add tests

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 committed:
Allow adding or replacing models through `create_deployment()`

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 committed:
Add usage examples for `create_deployment()`

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 committed:
Update docstring to tell users about the goal of `SageMakerDeploymentClient`

Signed-off-by: James Tran <tran.james2001@gmail.com>

dbczumar commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1116,12 +1133,17 @@ def _get_assumed_role_arn(**assume_role_credentials):
 def _assume_role_and_get_credentials(assume_role_arn):
     """
     Assume a new role in AWS and return the credentials for that role.
+    When ``assume_role_arn`` is ``None`` or an empty string,
+    this function does nothing and returns an empty dictionary.
 
     :param assume_role_arn: The ARN of the role that will be assumed
```
```suggestion
def _assume_role_and_get_credentials(assume_role_arn=None):
    """
    Assume a new role in AWS and return the credentials for that role.
    When ``assume_role_arn`` is ``None`` or an empty string,
    this function does nothing and returns an empty dictionary.

    :param assume_role_arn: Optional ARN of the role that will be assumed
```

dbczumar commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1064,6 +1053,34 @@ def _sigterm_handler(*_):
+    :return:
+    :rtype: str
+    """
+    help_str = """For detailed documentation on the SageMaker deployment client, please visit
+    https://mlflow.org/docs/latest/python_api/mlflow.sagemaker.html#mlflow.sagemaker.SageMakerDeploymentClient
```
Indentation nit:

```suggestion
    help_str = """\
    For detailed documentation on the SageMaker deployment client, please visit
    https://mlflow.org/docs/latest/python_api/mlflow.sagemaker.html#mlflow.sagemaker.SageMakerDeploymentClient
```

dbczumar commented on mlflow/sagemaker/__init__.py:
```diff
@@ -1747,6 +1769,704 @@ def _does_model_exist(model_name, sage_client):
+                message=(f"There was an error while retrieving the deployment: {exc}\n")
+            )
+
+    def predict(self, deployment_name, df):
+        raise NotImplementedError("This function is not implemented yet.")
+
+    def explain(self, deployment_name, df):  # pylint: disable=unused-argument
+        raise NotImplementedError("This function is not implemented yet.")
```
Can we remove these methods for now until they're implemented? This way, they won't show up in the docs.

dbczumar reviewed (approved):
@jamestran201 This is fantastic! I've tried it out and confirmed that roles are now being assumed for all operations and commandline data types are being parsed correctly. The docs render really well, and the examples are very helpful. Thank you for all your hard work here and for introducing thorough test coverage as well.

I've left 3 tiny documentation comments. I'm very excited to merge this once they are addressed. LGTM!

jamestran201 committed:
Fix indentation

Co-authored-by: Corey Zumar <39497902+dbczumar@users.noreply.github.com>
Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 committed:
Make parameter optional

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 committed:
Update docstring for unimplemented functions

Signed-off-by: James Tran <tran.james2001@gmail.com>

jamestran201 commented:
@dbczumar Nice! Thank you for your guidance!
